HW0/P4/P4.py

```diff
@@ -40,37 +62,86 @@
     # Read the observation array and plot it (Part 2)
     #####################
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
+    # References: http://stackoverflow.com/questions/4151128/what-are-the-differences-between-numpy-arrays-and-matrices-which-one-should-i-u
+    s_measured = np.loadtxt('P4_measurements.txt', delimiter=',')
+
+    F = np.zeros((3,3))
+    F[np.diag_indices(3)] = [1./rx, 1./ry, 1./rz]
+    m = np.dot(F,np.transpose(s_measured))
+
+    x_coords = m[0,:]
+    y_coords = m[1,:]
+    z_coords = m[2,:]
+
+    ax.plot(x_coords, y_coords, z_coords,
+            '.g', label='Observed trajectory')
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
-    # s = ?
+    # Propagation matrix
+    A = np.zeros((6,6))
+    A0, A1 = 1., 1 - c*dt
+    A[np.diag_indices(6)] = [1., 1., 1., A1, A1, A1]
+    A[0,3], A[1,4], A[2,5] = dt, dt, dt
+
+    # 'a' vector
+    a = np.zeros(6)
+    a[-1] = g*dt
+    a = np.transpose(a)
 
     # Initial conditions for s0
+    s0 = np.transpose( np.array( ([0, 0, 2, 15, 3.5, 4.0]) ) )
+
+    # 6xK 's' matrix
+    s = np.zeros( (6,K) )
+    s[:,0] = s0
+
     # Compute the rest of sk using Eq (1)
+    for cur_k in xrange(1, K):
+        s[:,cur_k] = np.dot(A,s[:,cur_k-1]) + a
+
+    x_coords = s[0,:]
+    y_coords = s[1,:]
+    z_coords = s[2,:]
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    ax.plot(x_coords, y_coords, z_coords,
+            '-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
-
     # Initial conditions for s0 and Sigma0
-    # Compute the rest of sk using Eqs (2), (3), (4), and (5)
+    sig_0 = 0.01 * np.eye(6)
+
+    B = np.zeros((6,6))
+    B[np.diag_indices(6)] = [bx, by, bz, bvx, bvy, bvz]
```

Comment:
np.diag_indices() is a good function. I'll try to use this function next time to avoid the long code block creating a 6 by 6 matrix with only diagonal entries nonzero.